libs/pika/async_mpi/include/pika/async_mpi/mpi_helpers.hpp

```diff
@@ -0,0 +1,208 @@
+#include <pika/execution/algorithms/then.hpp>
+#include <pika/execution/algorithms/transfer.hpp>
+#include <pika/execution_base/any_sender.hpp>
+#include <pika/execution_base/receiver.hpp>
+#include <pika/execution_base/sender.hpp>
+#include <pika/executors/thread_pool_scheduler.hpp>
+#include <pika/functional/detail/tag_fallback_invoke.hpp>
+#include <pika/functional/invoke.hpp>
+#include <pika/functional/invoke_fused.hpp>
+#include <pika/mpi_base/mpi.hpp>
+
+#include <exception>
+#include <type_traits>
+#include <utility>
+
+namespace pika::mpi::experimental::detail {
+
+    // -----------------------------------------------------------------
+    // by convention the title is 7 chars (for alignment)
+    using namespace pika::debug::detail;
+    template <int Level>
+    inline print_threshold<Level, 0> mpi_tran("MPITRAN");
+
+    namespace pud = pika::util::detail;
+    namespace exp = execution::experimental;
+
+    // -----------------------------------------------------------------
+    template <typename T>
+    struct any_sender_helper
+    {
+        using type = exp::unique_any_sender<T>;
+    };
+
+    template <>
+    struct any_sender_helper<void>
+    {
+        using type = exp::unique_any_sender<>;
+    };
+
+    // -----------------------------------------------------------------
+    // is func(Ts..., MPI_Request) invokable
+    template <typename F, typename... Ts>
+    inline constexpr bool is_mpi_request_invocable_v =
+        std::is_invocable_v<F, std::add_lvalue_reference_t<std::decay_t<Ts>>..., MPI_Request*>;
+
+    // -----------------------------------------------------------------
+    // get return type of func(Ts..., MPI_Request)
+    template <typename F, typename... Ts>
+    using mpi_request_invoke_result_t = std::decay_t<
+        std::invoke_result_t<F, std::add_lvalue_reference_t<std::decay_t<Ts>>..., MPI_Request*>>;
+
+    // -----------------------------------------------------------------
+    // return a scheduler on the mpi pool, with or without stack
+    inline auto mpi_pool_scheduler(bool stack = true)
+    {
+        if (stack)
+        {
+            return exp::thread_pool_scheduler{&resource::get_thread_pool(get_pool_name())};
+        }
+        else
+        {
+            return exp::with_stacksize(
+                exp::thread_pool_scheduler{&resource::get_thread_pool(get_pool_name())},
+                execution::thread_stacksize::nostack);
+        }
+    }
+
+    // -----------------------------------------------------------------
+    // return a scheduler on the default pool with added priority if requested
+    inline auto default_pool_scheduler(execution::thread_priority p)
+    {
+        if (p == execution::thread_priority::normal)
+        {
+            return exp::thread_pool_scheduler{&resource::get_thread_pool("default")};
+        }
+        return exp::with_priority(
+            exp::thread_pool_scheduler{&resource::get_thread_pool("default")}, p);
+    }
```

Comment:
This can stay here since it's in `detail`, but we could consider moving this to somewhere a bit more prominent (most likely resource partitioner/runtime or similar). No changes requested.